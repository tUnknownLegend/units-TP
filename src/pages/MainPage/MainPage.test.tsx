import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import { MainPage } from './MainPage';
import '@testing-library/jest-dom';
import { SortBy } from '../../types';

beforeAll(() => {
    jest.useFakeTimers().setSystemTime(new Date('2020-01-01T00:00:00'));
});

afterAll(() => {
    jest.useRealTimers();
});

afterEach(jest.clearAllMocks);

jest.mock('../../utils/getPrice', () => {
    return {
        __esModule: true,
        getPrice: jest.fn(() => '999 ₽'),
    };
});

jest.mock('../../utils/getNextSortBy', () => {
    return {
        __esModule: true,
        getNextSortBy: jest.fn(),
    };
});

jest.mock('../../utils/getNextSortBy', () => {
    return {
        __esModule: true,
        getNextSortBy: jest.fn((sortBy: SortBy) =>
            sortBy === 'по умолчанию'
                ? 'по возрастанию цены'
                : 'по убыванию цены'
        ),
    };
});

describe('checks text fields on main page', () => {
    it('should have VK Маркет name and match snapshot', () => {
        const renderedMainPage = render(<MainPage />);

        expect(renderedMainPage.asFragment()).toMatchSnapshot();
    });

    it('should change sort button innerText', () => {
        const renderedButton = render(<MainPage />);

        const sortButton = renderedButton.getByTestId('sort-button-id');

        fireEvent.click(sortButton);
        expect(sortButton.textContent).toStrictEqual(
            'Сортировать по возрастанию цены'
        );
        fireEvent.click(sortButton);
        expect(sortButton.textContent).toStrictEqual(
            'Сортировать по убыванию цены'
        );
    });
});

describe('Main page select category', () => {
    it('click on the first matched button', () => {
        const renderedMainPage = render(<MainPage />);

        const categoryElement = renderedMainPage.getAllByTestId(
            'button-select-category'
        )[0];

        expect(categoryElement).not.toBeUndefined();
        fireEvent.click(categoryElement);

        const productEdited = renderedMainPage.getAllByTestId(
            'product-card__category'
        );

        expect(
            productEdited.filter(
                (productCard) =>
                    productCard.innerText === categoryElement.innerText
            ).length
        ).toEqual(productEdited.length);
    });
});
