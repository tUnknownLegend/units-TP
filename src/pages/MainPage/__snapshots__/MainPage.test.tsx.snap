// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checks text fields on main page should have VK Маркет name and match snapshot 1`] = `
<DocumentFragment>
  <div
    class="main-page"
  >
    <h2
      class="main-page__title"
    >
      VK Маркет
    </h2>
    <h3>
      00:00:00
    </h3>
    <div
      class="main-page__parameters"
    >
      <div
        class="categories"
      >
        <div
          class="categories__badge"
          data-testid="button-select-category"
        >
          Одежда
        </div>
        <div
          class="categories__badge"
          data-testid="button-select-category"
        >
          Для дома
        </div>
        <div
          class="categories__badge"
          data-testid="button-select-category"
        >
          Электроника
        </div>
      </div>
      <button
        class="sort-button"
        data-testid="sort-button-id"
      >
        Сортировать по умолчанию
      </button>
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          IPhone 14 Pro
        </h2>
        <div
          class="product-card__description"
        >
          Latest iphone, buy it now
        </div>
        <h1
          class="product-card__price"
        >
          999 ₽
        </h1>
        <div
          class="product-card__category"
          data-testid="product-card__category"
        >
          Электроника
        </div>
      </div>
      <img
        alt="IPhone 14 Pro"
        class="product-card__image"
        data-testid="product-card__image"
        src="/iphone.png"
      />
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Костюм гуся
        </h2>
        <div
          class="product-card__description"
        >
          Запускаем гуся, работяги
        </div>
        <h1
          class="product-card__price"
        >
          999 ₽
        </h1>
        <div
          class="product-card__category"
          data-testid="product-card__category"
        >
          Одежда
        </div>
      </div>
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Настольная лампа
        </h2>
        <div
          class="product-card__description"
        >
          Говорят, что ее использовали в pixar
        </div>
        <h1
          class="product-card__price"
        >
          999 ₽
        </h1>
        <div
          class="product-card__category"
          data-testid="product-card__category"
        >
          Для дома
        </div>
      </div>
      <img
        alt="Настольная лампа"
        class="product-card__image"
        data-testid="product-card__image"
        src="/lamp.png"
      />
    </div>
    <div
      class="product-card"
    >
      <div
        class="product-card__text"
      >
        <h2
          class="product-card__name"
        >
          Принтер
        </h2>
        <div
          class="product-card__description"
        >
          Незаменимая вещь для студента
        </div>
        <h1
          class="product-card__price"
        >
          999 ₽
        </h1>
        <div
          class="product-card__category"
          data-testid="product-card__category"
        >
          Электроника
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
